### ISSUE TITLE
SourceNode.FromNode will trigger unnecessary creation of AnnotationLists.

### ISSUE THREAD

**Issue by ewoutkramer** [FirelyTeam/firely-net-sdk#1913]
**Describe the bug**
In #1912, we devised a way to avoid creating empty lists of Annotations.  Unfortunately, `FromNode()` shows how easy it is to still trigger unnecessary allocations:

This simple line in `buildNode()` in `SourceNode.cs`:

```csharp
   var rts = node.Annotation<IResourceTypeSupplier>();
```

will trigger creation of the Lazy<T>.  This can still quite easily be solved, by making sure that the implementation in `SourceNode.cs` of `IAnnotated.Annotation()` will not unneccessarily trigger the creation of the Lazy<T>.

The next bit in `buildNode()` shows a comparable issue:

```csharp
  foreach (var t in annotationsToCopy ?? Enumerable.Empty<Type>())
                foreach (var ann in node.Annotations(t))
                    me.AddAnnotation(ann);
```

Currently, `node.Annotations(t)` will trigger creation of an annotationlist, and so will  `AddAnnotation()`.

**Expected behavior**
We should try to avoid allocations as much as possible, starting with the code in `DomNode.cs` and `SourceNode.cs`. Time permitting, we could review the other node/typedelement code in the SDK for similar issues.

**Comment by marcovisserFurore** [comment]
This following line in `buildNode()` in `SourceNode.cs` does _not_ trigger the creation of the Lazy:

```c#
   var rts = node.Annotation<IResourceTypeSupplier>();
```

The extension methode `Annotation<T>(this ISourceNode nav)` will trigger the method `Annotation<A>(this IAnnotated annotated)`, where `annotated` is here the `SourceNode`. The implementation of this method in `SourceNode` looks like this:

```c# 
        public IEnumerable<object> Annotations(Type type)
        {
            return type == typeof(SourceNode) || type == typeof(ISourceNode) || type == typeof(IResourceTypeSupplier)
                ? (new[] { this })
                : AnnotationsInternal.OfType(type);
        }
``` 
So, when `type` is an `IResourceTypeSupplier` then it will return `this` and _not_ the `AnnotationsInternal`.

In other words the AnnotationList will stull not be initialized, which is good thing.

**Comment by marcovisserFurore** [comment]
In the above code snippet, I changed the following line:

```c#
  : AnnotationsInternal.OfType(type);
```

to

```c#
  HasAnnotations ? AnnotationsInternal.OfType(type) : Enumerable.Empty<object>();
````
which will not create the AnnotationList. 

See also PR https://github.com/FirelyTeam/firely-net-common/pull/171
